File: template-service/src/test/java/com/sensa/templateservice/service/TemplateServiceTest.java
package com.sensa.templateservice.service;

import com.sensa.templateservice.dto.TemplateRequest;
import com.sensa.templateservice.dto.TemplateResponse;
import com.sensa.templateservice.entity.Template;
import com.sensa.templateservice.exception.TemplateAlreadyExistException;
import com.sensa.templateservice.exception.TemplateNotFoundException;
import com.sensa.templateservice.mapper.TemplateMapper;
import com.sensa.templateservice.repository.TemplateRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class TemplateServiceTest {

    @Mock
    private TemplateRepository templateRepository;

    @Mock
    private TemplateMapper templateMapper;

    @InjectMocks
    private TemplateService templateService;

    @Test
    void testCreateTemplateSuccess() {
        String username = "user1";
        TemplateRequest request = TemplateRequest.builder()
                .clientUsername(username)
                .title("Template1")
                .content("Content1")
                .build();

        // Сценарий: шаблон с таким заголовком не существует
        when(templateRepository.existTemplateByUsernameAndTitle(username, request.title())).thenReturn(false);

        // Маппинг DTO в сущность
        Template templateEntity = Template.builder()
                .clientUsername(null)
                .title(request.title())
                .content(request.content())
                .build();
        when(templateMapper.mapToEntity(request)).thenReturn(templateEntity);

        // Предположим, что метод addUsername устанавливает username в сущность и возвращает её
        Template updatedTemplate = templateEntity.addUsername(username);

        // При сохранении возвращается обновлённая сущность
        when(templateRepository.save(updatedTemplate)).thenReturn(updatedTemplate);

        // Маппер преобразует сущность в ответ
        TemplateResponse response = TemplateResponse.builder()
                .clientUsername(username)
                .title(request.title())
                .content(request.content())
                .build();
        when(templateMapper.mapToResponse(updatedTemplate)).thenReturn(response);

        TemplateResponse result = templateService.createTemplate(username, request);
        assertEquals(response, result);
    }

    @Test
    void testCreateTemplateAlreadyExists() {
        String username = "user1";
        TemplateRequest request = TemplateRequest.builder()
                .clientUsername(username)
                .title("Template1")
                .content("Content1")
                .build();

        // Сценарий: шаблон уже существует
        when(templateRepository.existTemplateByUsernameAndTitle(username, request.title())).thenReturn(true);

        TemplateAlreadyExistException exception = assertThrows(TemplateAlreadyExistException.class,
                () -> templateService.createTemplate(username, request));
        String expectedMessage = String.format("Template with title %s already exists for user: %s", request.title(), username);
        assertEquals(expectedMessage, exception.getMessage());
    }

    @Test
    void testFindTemplateSuccess() {
        String username = "user1";
        String title = "Template1";
        Template templateEntity = Template.builder()
                .clientUsername(username)
                .title(title)
                .content("Content1")
                .build();

        when(templateRepository.findByClientUsernameAndTitle(username, title)).thenReturn(Optional.of(templateEntity));

        TemplateResponse response = TemplateResponse.builder()
                .clientUsername(username)
                .title(title)
                .content("Content1")
                .build();
        when(templateMapper.mapToResponse(templateEntity)).thenReturn(response);

        TemplateResponse result = templateService.findTemplate(username, title);
        assertEquals(response, result);
    }

    @Test
    void testFindTemplateNotFound() {
        String username = "user1";
        String title = "Template1";
        when(templateRepository.findByClientUsernameAndTitle(username, title)).thenReturn(Optional.empty());

        TemplateNotFoundException exception = assertThrows(TemplateNotFoundException.class,
                () -> templateService.findTemplate(username, title));
        String expectedMessage = String.format("Template with title %s not found", title);
        assertEquals(expectedMessage, exception.getMessage());
    }

    @Test
    void testUpdateTemplateSuccess() {
        String username = "user1";
        String oldTitle = "Template1";
        TemplateRequest request = TemplateRequest.builder()
                .clientUsername(username)
                .title("UpdatedTemplate")
                .content("UpdatedContent")
                .build();

        Template existingTemplate = Template.builder()
                .clientUsername(username)
                .title(oldTitle)
                .content("OldContent")
                .build();

        when(templateRepository.findByClientUsernameAndTitle(username, oldTitle)).thenReturn(Optional.of(existingTemplate));

        // Обновляем поля: в сервисе вызывается Optional.ofNullable(...).ifPresent(...)
        existingTemplate.setTitle(request.title());
        existingTemplate.setContent(request.content());

        when(templateRepository.save(existingTemplate)).thenReturn(existingTemplate);

        TemplateResponse response = TemplateResponse.builder()
                .clientUsername(username)
                .title(request.title())
                .content(request.content())
                .build();
        when(templateMapper.mapToResponse(existingTemplate)).thenReturn(response);

        TemplateResponse result = templateService.updateTemplate(username, oldTitle, request);
        assertEquals(response, result);
    }

    @Test
    void testUpdateTemplateNotFound() {
        String username = "user1";
        String oldTitle = "Template1";
        TemplateRequest request = TemplateRequest.builder()
                .clientUsername(username)
                .title("NewTitle")
                .content("NewContent")
                .build();

        when(templateRepository.findByClientUsernameAndTitle(username, oldTitle)).thenReturn(Optional.empty());

        TemplateNotFoundException exception = assertThrows(TemplateNotFoundException.class,
                () -> templateService.updateTemplate(username, oldTitle, request));
        String expectedMessage = String.format("Template with title %s not found for user: %s", request.title(), username);
        assertEquals(expectedMessage, exception.getMessage());
    }

    @Test
    void testDeleteTemplateSuccess() {
        String username = "user1";
        String title = "Template1";
        when(templateRepository.deleteByClientUsernameAndTitle(username, title)).thenReturn(1);

        Boolean result = templateService.deleteTemplate(username, title);
        assertTrue(result);
    }

    @Test
    void testDeleteTemplateFailure() {
        String username = "user1";
        String title = "Template1";
        when(templateRepository.deleteByClientUsernameAndTitle(username, title)).thenReturn(0);

        Boolean result = templateService.deleteTemplate(username, title);
        assertFalse(result);
    }
}

File: template-service/src/test/kotlin/com/sensa/templateservice/service/TemplateServiceTest.kt
package com.sensa.templateservice.service

import com.sensa.templateservice.dto.TemplateRequest
import com.sensa.templateservice.dto.TemplateResponse
import com.sensa.templateservice.entity.Template
import com.sensa.templateservice.entity.TemplateChannel
import com.sensa.templateservice.exception.TemplateAlreadyExistsException
import com.sensa.templateservice.exception.TemplateNotFoundException
import com.sensa.templateservice.mapper.TemplateMapper
import com.sensa.templateservice.repository.TemplateRepository
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import java.time.LocalDateTime
import java.util.UUID

class TemplateServiceTest {

    private lateinit var templateRepository: TemplateRepository
    private lateinit var templateMapper: TemplateMapper
    private lateinit var templateService: TemplateService

    private val userId = UUID.randomUUID()
    private val now = LocalDateTime.now()

    @BeforeEach
    fun setUp() {
        templateRepository = Mockito.mock(TemplateRepository::class.java)
        templateMapper = Mockito.mock(TemplateMapper::class.java)
        templateService = TemplateService(templateRepository, templateMapper)
    }

    private fun buildEntity(
        id: Long = 1L,
        templateName: String = "alert-template",
        description: String? = "Emergency alert",
        channel: TemplateChannel = TemplateChannel.PUSH,
        content: String = "Alert: {{message}}",
        createdBy: UUID = userId
    ) = Template(
        id = id,
        templateName = templateName,
        description = description,
        channel = channel,
        content = content,
        createdBy = createdBy,
        createdAt = now,
        updatedAt = now
    )

    private fun buildRequest(
        templateName: String = "alert-template",
        description: String? = "Emergency alert",
        channel: TemplateChannel = TemplateChannel.PUSH,
        content: String = "Alert: {{message}}"
    ) = TemplateRequest(
        templateName = templateName,
        description = description,
        channel = channel,
        content = content
    )

    private fun buildResponse(
        id: Long = 1L,
        templateName: String = "alert-template",
        description: String? = "Emergency alert",
        channel: TemplateChannel = TemplateChannel.PUSH,
        content: String = "Alert: {{message}}"
    ) = TemplateResponse(
        id = id,
        templateName = templateName,
        description = description,
        channel = channel,
        content = content,
        createdBy = userId,
        createdAt = now,
        updatedAt = now
    )

    @Test
    fun `create template success`() {
        val request = buildRequest()
        val entity = buildEntity()
        val response = buildResponse()

        Mockito.`when`(templateRepository.existsByTemplateName(request.templateName)).thenReturn(false)
        Mockito.`when`(templateMapper.toEntity(request, userId)).thenReturn(entity)
        Mockito.`when`(templateRepository.save(entity)).thenReturn(entity)
        Mockito.`when`(templateMapper.toResponse(entity)).thenReturn(response)

        val result = templateService.create(request, userId)

        assertNotNull(result)
        assertEquals("alert-template", result.templateName)
        Mockito.verify(templateRepository).save(entity)
    }

    @Test
    fun `create template already exists throws exception`() {
        val request = buildRequest()
        Mockito.`when`(templateRepository.existsByTemplateName(request.templateName)).thenReturn(true)

        assertThrows(TemplateAlreadyExistsException::class.java) {
            templateService.create(request, userId)
        }
    }

    @Test
    fun `get by template name success`() {
        val entity = buildEntity()
        val response = buildResponse()

        Mockito.`when`(templateRepository.findByTemplateName("alert-template")).thenReturn(entity)
        Mockito.`when`(templateMapper.toResponse(entity)).thenReturn(response)

        val result = templateService.getByTemplateName("alert-template")

        assertEquals("alert-template", result.templateName)
    }

    @Test
    fun `get by template name not found throws exception`() {
        Mockito.`when`(templateRepository.findByTemplateName("unknown")).thenReturn(null)

        assertThrows(TemplateNotFoundException::class.java) {
            templateService.getByTemplateName("unknown")
        }
    }

    @Test
    fun `get all returns list`() {
        val entity = buildEntity()
        val response = buildResponse()

        Mockito.`when`(templateRepository.findAll()).thenReturn(listOf(entity))
        Mockito.`when`(templateMapper.toResponse(entity)).thenReturn(response)

        val result = templateService.getAll()

        assertEquals(1, result.size)
    }

    @Test
    fun `get by channel returns filtered list`() {
        val entity = buildEntity()
        val response = buildResponse()

        Mockito.`when`(templateRepository.findByChannel(TemplateChannel.PUSH)).thenReturn(listOf(entity))
        Mockito.`when`(templateMapper.toResponse(entity)).thenReturn(response)

        val result = templateService.getByChannel(TemplateChannel.PUSH)

        assertEquals(1, result.size)
    }

    @Test
    fun `update template success`() {
        val existing = buildEntity()
        val request = buildRequest(templateName = "updated-template")
        val response = buildResponse(templateName = "updated-template")

        Mockito.`when`(templateRepository.findByTemplateName("alert-template")).thenReturn(existing)
        Mockito.`when`(templateRepository.save(existing)).thenReturn(existing)
        Mockito.`when`(templateMapper.toResponse(existing)).thenReturn(response)

        val result = templateService.update("alert-template", request)

        assertNotNull(result)
        assertEquals("updated-template", result.templateName)
        Mockito.verify(templateMapper).updateEntity(existing, request)
    }

    @Test
    fun `update template not found throws exception`() {
        val request = buildRequest()
        Mockito.`when`(templateRepository.findByTemplateName("unknown")).thenReturn(null)

        assertThrows(TemplateNotFoundException::class.java) {
            templateService.update("unknown", request)
        }
    }

    @Test
    fun `delete template success`() {
        Mockito.`when`(templateRepository.deleteByTemplateName("alert-template")).thenReturn(1)

        assertDoesNotThrow { templateService.delete("alert-template") }
    }

    @Test
    fun `delete template not found throws exception`() {
        Mockito.`when`(templateRepository.deleteByTemplateName("unknown")).thenReturn(0)

        assertThrows(TemplateNotFoundException::class.java) {
            templateService.delete("unknown")
        }
    }
}
